lib/virtio_mmio/virtio_mmio_hvl.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2022 Wind River Systems, Inc.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * $FreeBSD$
+ */
+
+#if defined(HVL_VIRTIO)
+
+#include <openamp/open_amp.h>
+#include <openamp/virtqueue.h>
+#include <openamp/virtio.h>
+#include <openamp/virtio_mmio.h>
+#include <openamp/virtio_mmio_hvl.h>
+#include <metal/device.h>
+#include <metal/mutex.h>
+#include <metal/sleep.h>
+
+/*
+ * The implementation depends on OS-specific services for memory pool management
+ * routines, blocking wait and IPI.
+ * TODO: Replace them in a future iteration of the hypervisorless virtio feature
+ * set with APIs (to be introduced) in libmetal.
+ */
+
+/* hypervisorless virtio callback framework */
+static virtio_mmio_hvl_cb_t _cb_funcs[VIRTIO_MMIO_HVL_CB_MAX];
+static void *_cb_args[VIRTIO_MMIO_HVL_CB_MAX];
+static int _num_cb_funcs;
+
+/* pre-shared memory */
+uint32_t __shmem0_pool_start __attribute__((weak));
+uint32_t __shmem0_end __attribute__((weak));
+
+void virtio_mmio_hvl_init(struct virtio_device *vdev)
+{
+	static int init;
+	uint32_t *shmem_pool_mem = NULL;
+	uint32_t shmem_pool_size = 0;
+	struct virtio_mmio_device *vmdev = metal_container_of(vdev,
+							      struct virtio_mmio_device, vdev);
+
+	VIRTIO_ASSERT(vmdev->hvl_func->hvl_shm_pool_init,
+		      "Invalid HVL dispatch configuration");
+	VIRTIO_ASSERT(vmdev->hvl_func->hvl_shm_pool_alloc,
+		      "Invalid HVL dispatch configuration");
+	VIRTIO_ASSERT(vmdev->hvl_func->hvl_shm_pool_free,
+		      "Invalid HVL dispatch configuration");
+	VIRTIO_ASSERT(vmdev->hvl_func->hvl_wait_cfg_event,
+		      "Invalid HVL dispatch configuration");
+	VIRTIO_ASSERT(vmdev->hvl_func->hvl_ipi,
+		      "Invalid HVL dispatch configuration");
+
+	if (!init) {
+		shmem_pool_mem = &__shmem0_pool_start;
+		if (shmem_pool_mem != 0) {
+			shmem_pool_size = ((uintptr_t)&__shmem0_end -
+					   (uintptr_t)&__shmem0_pool_start);
+			vmdev->hvl_func->hvl_shm_pool_init(shmem_pool_mem, shmem_pool_size);
+			init = 1;
+		}
+	}
+	if (vdev) {
+		virtio_mmio_hvl_set_ipi(vdev, vmdev->hvl_func->hvl_ipi, NULL);
+	}
+	metal_mutex_init(&vmdev->cfg_sem);
+	metal_mutex_acquire(&vmdev->cfg_sem);
```

Comment:
This will not block. It's just a workaround for the Zephyr mutex API in libmetal which does a k_sem_init(m, 1, 1), i.e. "sem full". The call to acquire is meant to "empty" it.